File: ts/src/relay/routes/read.ts
import { Hono } from 'hono';
import type { RelayState } from '../app';

export function createReadRoutes(state: RelayState) {
  const app = new Hono();

  // GET /rad/visible/:filePath - visible な write 一覧
  app.get('/rad/visible/:filePath', (c) => {
    const filePath = c.req.param('filePath');
    const ops = state.oplog.getAllOperations();

    // filePath に関連する visible な write 操作を抽出
    const visibleWrites = ops.filter(op =>
      op.status === 'visible' &&
      op.type === 'write' &&
      op.regionId.startsWith(filePath + ':')
    );

    return c.json(visibleWrites, 200);
  });

  // GET /rad/files/:filePath - accepted なファイル内容
  app.get('/rad/files/:filePath', (c) => {
    const filePath = c.req.param('filePath');
    const ops = state.oplog.getAllOperations();

    // filePath に関連する accepted な write 操作を抽出
    const acceptedWrites = ops.filter(op =>
      op.status === 'accepted' &&
      op.type === 'write' &&
      op.regionId.startsWith(filePath + ':')
    );

    // 最新の内容を返す（最後の accepted write）
    if (acceptedWrites.length > 0) {
      const latest = acceptedWrites[acceptedWrites.length - 1];
      return c.text(latest.content, 200);
    }

    return c.json({ error: 'file not found or no accepted writes' }, 404);
  });

  // GET /rad/files - ファイル一覧
  app.get('/rad/files', (c) => {
    const ops = state.oplog.getAllOperations();
    const files = new Set<string>();

    ops.forEach(op => {
      if (op.type === 'write') {
        const filePath = op.regionId.split(':')[0];
        files.add(filePath);
      }
    });

    return c.json(Array.from(files), 200);
  });

  // GET /rad/regions/:filePath - コード領域一覧
  app.get('/rad/regions/:filePath', (c) => {
    const filePath = c.req.param('filePath');
    const regions = state.regionMap.getAllRegions();

    // filePath に関連する領域を抽出
    const fileRegions = regions.filter(r => r.filePath === filePath);

    return c.json(fileRegions, 200);
  });

  // GET /rad/log - 操作ログ
  app.get('/rad/log', (c) => {
    const regionId = c.req.query('regionId');
    const participantId = c.req.query('participantId');
    const since = c.req.query('since');
    const limit = c.req.query('limit');

    let ops = state.oplog.getAllOperations();

    // フィルタリング
    if (regionId) {
      ops = ops.filter(op => op.regionId === regionId);
    }
    if (participantId) {
      ops = ops.filter(op => op.participantId === participantId);
    }
    if (since) {
      const sinceTs = parseInt(since);
      ops = ops.filter(op => op.timestamp >= sinceTs);
    }
    if (limit) {
      const limitNum = parseInt(limit);
      ops = ops.slice(0, limitNum);
    }

    return c.json(ops, 200);
  });

  // POST /rad/compact - コンパクション実行
  app.post('/rad/compact', (c) => {
    // RP9 では簡易実装（accepted 操作を削除するのみ）
    // RP12 で永続化と合わせて完全実装
    const ops = state.oplog.getAllOperations();
    const acceptedOps = ops.filter(op => op.status === 'accepted');
    const nonAcceptedOps = ops.filter(op => op.status !== 'accepted');

    // accepted でない操作のみを残す
    state.oplog.loadOperations(nonAcceptedOps);

    return c.json({
      compacted: acceptedOps.length,
      message: 'compaction completed'
    }, 200);
  });

  // POST /rad/sync/git - Git 双方向同期（RP11 で実装）
  app.post('/rad/sync/git', (c) => {
    return c.json({ error: 'not implemented yet (RP11)' }, 501);
  });

  return app;
}

File: rust/src/relay/read.rs
use axum::{Json, extract::{State, Path, Query}, http::StatusCode};
use serde::{Deserialize, Serialize};
use super::state::SharedState;
use super::participants::ErrorResponse;
use crate::types::{Operation, CodeRegion, OpType};
use std::collections::HashSet;

#[derive(Debug, Deserialize)]
pub struct LogQuery {
    #[serde(rename = "regionId")]
    region_id: Option<String>,
    #[serde(rename = "participantId")]
    participant_id: Option<String>,
    since: Option<u64>,
    limit: Option<usize>,
}

#[derive(Debug, Serialize)]
pub struct CompactResponse {
    compacted: usize,
    message: String,
}

// GET /rad/visible/:path
pub async fn visible(
    State(state): State<SharedState>,
    Path(file_path): Path<String>,
) -> Json<Vec<Operation>> {
    let oplog = state.oplog.lock().unwrap();
    let ops = oplog.all();

    let visible_writes: Vec<Operation> = ops
        .into_iter()
        .filter(|op| {
            format!("{:?}", op.status).to_lowercase() == "visible"
                && op.op_type == OpType::Write
                && op.region_id.starts_with(&format!("{}:", file_path))
        })
        .cloned()
        .collect();

    Json(visible_writes)
}

// GET /rad/files/:path
pub async fn file(
    State(state): State<SharedState>,
    Path(file_path): Path<String>,
) -> Result<String, (StatusCode, Json<ErrorResponse>)> {
    let oplog = state.oplog.lock().unwrap();
    let ops = oplog.all();

    let accepted_writes: Vec<Operation> = ops
        .into_iter()
        .filter(|op| {
            format!("{:?}", op.status).to_lowercase() == "accepted"
                && op.op_type == OpType::Write
                && op.region_id.starts_with(&format!("{}:", file_path))
        })
        .cloned()
        .collect();

    if let Some(latest) = accepted_writes.last() {
        Ok(latest.content.clone())
    } else {
        Err((
            StatusCode::NOT_FOUND,
            Json(ErrorResponse {
                error: "file not found or no accepted writes".to_string(),
            }),
        ))
    }
}

// GET /rad/files
pub async fn file_list(State(state): State<SharedState>) -> Json<Vec<String>> {
    let oplog = state.oplog.lock().unwrap();
    let ops = oplog.all();

    let mut files = HashSet::new();
    for op in ops {
        if op.op_type == OpType::Write {
            if let Some(colon_pos) = op.region_id.find(':') {
                files.insert(op.region_id[..colon_pos].to_string());
            }
        }
    }

    Json(files.into_iter().collect())
}

// GET /rad/regions/:path
pub async fn regions(
    State(state): State<SharedState>,
    Path(file_path): Path<String>,
) -> Json<Vec<CodeRegion>> {
    let region_map = state.region_map.lock().unwrap();
    let all_regions = region_map.get_all_regions();

    let file_regions: Vec<CodeRegion> = all_regions
        .into_iter()
        .filter(|r| r.file_path == file_path)
        .collect();

    Json(file_regions)
}

// GET /rad/log
pub async fn log(
    State(state): State<SharedState>,
    Query(query): Query<LogQuery>,
) -> Json<Vec<Operation>> {
    let oplog = state.oplog.lock().unwrap();
    let mut ops: Vec<Operation> = oplog.all().to_vec();

    // フィルタリング
    if let Some(ref region_id) = query.region_id {
        ops.retain(|op| &op.region_id == region_id);
    }
    if let Some(ref participant_id) = query.participant_id {
        ops.retain(|op| &op.participant_id == participant_id);
    }
    if let Some(since) = query.since {
        ops.retain(|op| op.timestamp >= since);
    }
    if let Some(limit) = query.limit {
        ops.truncate(limit);
    }

    Json(ops)
}

// POST /rad/compact
pub async fn compact(State(state): State<SharedState>) -> Json<CompactResponse> {
    let mut oplog = state.oplog.lock().unwrap();
    let ops = oplog.all().to_vec();

    let accepted_ops: Vec<Operation> = ops
        .iter()
        .filter(|op| format!("{:?}", op.status).to_lowercase() == "accepted")
        .cloned()
        .collect();

    let non_accepted_ops: Vec<Operation> = ops
        .into_iter()
        .filter(|op| format!("{:?}", op.status).to_lowercase() != "accepted")
        .collect();

    let compacted_count = accepted_ops.len();

    // accepted でない操作のみを残す
    // OpLog を新しいものに置き換える（load_operationsの代わり）
    *oplog = crate::oplog::OpLog::new();
    for op in non_accepted_ops {
        oplog.append(op);
    }

    Json(CompactResponse {
        compacted: compacted_count,
        message: "compaction completed".to_string(),
    })
}

// POST /rad/sync/git
pub async fn sync_git() -> (StatusCode, Json<ErrorResponse>) {
    (
        StatusCode::NOT_IMPLEMENTED,
        Json(ErrorResponse {
            error: "not implemented yet (RP11)".to_string(),
        }),
    )
}
